Conch/lib/Test/Conch/Validation.pm

```diff
@@ -0,0 +1,285 @@
+
+This function performs the following tests:
+
+=over
+
+=item
+
+Test whether the validation builds.
+
+=item
+
+Tests whether the validations defines the required C<name>, C<version>,
+and C<description> attributes.
+
+=item
+
+For each case, test whether the validation dies if C<< dies => 1 >> is specified in
+the case, or lives and produces the expected number of successful and failing
+validation results.
+
+=back
+
+The required arguments are the Conch Validation module as a string, keyword
+arguments specifying optional models available to the Validation, and a keyword
+argument specifying the cases for the test to use.
+
+The available models are C<hardware_product>, C<device_location>,
+C<device_settings>, and C<device>. Their attributes are defined with a hashref,
+which will be constructed to the correct classes in the body of
+L<test_validation>. For example:
+
+	test_validation(
+		'Conch::Validation::TestValidation',
+		hardware_product => {
+			name => 'Product Name',
+			vendor => 'Product Vendor',
+			profile => {
+				num_cpu => 2
+			}
+		},
+		device_location => {
+			rack_unit       => 2,
+			datacenter_rack => { slots => [ 1, 2, 3 ] }
+		},
+		device_settings => {
+			foo => 'bar'
+		},
+		device => {
+			triton_setup => 1,
+		},
+
+		cases => [ ... ]
+	);
+
+C<cases> is a list of hashrefs defining each of the test cases. Each case
+specifies the input data and attributes representing the expected results. Each
+test case may raise an error (die) or may produce 0 or more validation results.
+A test case is specified with a hashref with the attributes:
+
+=over
+
+=item C<data>
+
+A hashref of the input data provide to the Validation. An empty hashref will be provided by default.
+
+=item C<dies>
+
+A boolean attribute (0 or 1) to mark that this test case is expected to die
+during validation. A validatino may die due to schema errors, C<< $self->die >>
+or C<die> called in the Validation logic, or other code errors.  Defaults to 0.
+
+=item C<success_num>
+
+The number of expected successful validation results from running the
+Validation with the provided C<data>. Defaults to 0.
+
+=item C<failure_num>
+
+The number of expected failing validation results from running the Validation
+with the provided C<data>. Defaults to 0
+
+=item C<description>
+
+Optional description of the test case. Provides documentation and adds the
+description to test failure messages.
+
+=item C<debug>
+
+Optional boolean flag to provide additional diagnostic information when running
+the case using L<Test::More::diag>. This is helpful during development of test
+cases, but should be removed before committing.
+
+=back
+
+Example:
+
+	test_validation(
+		'Conch::Validation::TestValidation',
+		cases => [
+			{
+				data        => { hello => 'world' },
+				dies        => 1,
+				success_num => 3,
+				failure_num => 3,
+				description => 'Hello world test case',
+				debug       => 1
+			}
+		]
+	);
+
+
+=cut
+
+sub test_validation {
```

Comment:
This is really hard to follow with all the diags and perltidy's odd manner of formatting long lines.